engine/src/main/java/org/hibernate/validator/engine/HibernateConstraintViolation.java

```diff
@@ -0,0 +1,21 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.engine;
+
+import javax.validation.ConstraintViolation;
+
+/**
+ * A custom {@link ConstraintViolation} which allows to get additional information for a constraint violation.
+ *
+ * @since 5.3
+ */
+public interface HibernateConstraintViolation<T> extends ConstraintViolation<T> {
+	/**
+	 * @return an object that further describes the violation or {@code null}.
+	 */
+	Object getViolationContext();
```

Comment:
What do you think about:

    public <C> C getViolationContext(Class<C> type) 

This would potentially avoid a cast for the user. The question would be what to do if there is a context object, but of different type. I guess one would need to return `null`, since throwing an exception seems wrong. The user might use multiple custom constraints with different context objects.